Decorators less verbose and subscriptions

playerx opened the pull request:
- Resolved https://github.com/indigotech/graphql-schema-decorator/issues/12 - Making decorators less verbose
- Added Subscriptions

Commit:
fixed typings

Commit:
created generic description method & field decorator can set description

Commit:
implemented NonNull as Field option

Commit:
implemente4d description as schema option

Commit:
implemented description for Query & Mutation options

Commit:
created description for List option

Commit:
made option optional

Commit:
completed implementing description as part of other decoration option

Commit:
refactored setDescriptionMetadata, simplified

playerx commented:
How to use:
```
@ObjectType({ description: 'A user type.' })
class User {
    @Field({ type: graphql.GraphQLID }) id: string;
    @Field() name: string;
    @Field() email: string;
}
```
```
@InputObjectType({ description: 'Test 123123' })
class UserForUpdate {
    @Field() name: string;
    @Field() email: string;
}
```

thiago-soliveira commented on src/decorator.ts:
```diff
@@ -66,13 +66,41 @@ export interface EnumValueMetadata {
     type?: any;
 }
 
+export interface ListOption {
```
Since all these `description` fileds mean the same thing and it is part of every available option object, what if you'd extract them to an interface and then you'd extend that interface from all those options?  

Something like:
```typescript
export interface DefaultOption {
  description?: string;
}

export interface ObjectTypeOption extends DefaultOption { }
```

playerx replied:
Yes I thought about that also but kept it for next step, I was thinking about refactoring current code base, for now everything (about decorators) is in one file and maybe it would be good to split it in separate files, file per decorator + 1 common file.

thiago-soliveira commented on src/decorator.ts:
```diff
@@ -66,13 +66,41 @@ export interface EnumValueMetadata {
 
 export interface FieldOption {
     type?: any;
+    description?: string;
```
What about the pagination? It would be nice to have it here as well.

playerx replied:
Yes I agree, but first I've implemented just description field to get your feedback and pagination can be implemented with the same way later.

thiago-soliveira commented:
Two more things apart from that
- Could you please write some tests covering this new scenario?
- You'd be really nice if could mention this new approach on docs.

playerx commented:
Okey, I'll need some time and I'll make those things

playerx committed:
created new sample

playerx committed:
removed unused decorators

playerx committed:
formatted

playerx committed:
created DefaultOption

playerx committed:
created new options for field decorator & created tests

playerx committed:
updated sample

playerx committed:
added:
* description option for @EnumType
* description option for @Value
* nonNull option for @Arg

updated tests

playerx committed:
updated documentation & samples

playerx commented:
- [x] covered new scenarios with tests
- [x] added pagination, list options in Field decorator
- [x] added description option for Enum and Value decorators
- [x] updated documentation
- [x] created new sample project using Apollo server

@thiago-soliveira  check out latest updates please

playerx commented:
Also its important your feedback about new sample project structure

playerx committed:
created initial subscription support

playerx committed:
* removed debug info
* completed subscription implementation
* updated graphql up to 0.11.7

playerx committed:
updated demo project, crated subscription sample with documentation

playerx commented:
We have cool updates, I've added subscription implementation, and created demo in Apollo-server example

playerx committed:
fixed undefined parameter

playerx committed:
updated types for graphql up to 0.11.5

playerx committed:
updated example project, created paging & orderby sample

playerx committed:
* refactoring
* added hapi server

playerx committed:
updated documentation

playerx committed:
added printSchema endpoint

playerx committed:
added esnext lib

playerx committed:
fixed examples

thiago-soliveira commented on src/schema_factory.ts:
```diff
@@ -31,53 +64,28 @@ export function schemaFactory(target: Function) {
-        const fields: {[key: string]: any} = {};
-        let mutationTypeFn: Function;
-        let fieldMetadataList: FieldTypeMetadata[];
+    // let result: any = {
```
Could you remove this code that is commented out?

playerx replied:
done

thiago-soliveira reviewed (approved):
Very nice changes. The server example will help a lot on development and for users as an example.
+1 for the subscription as well.
Let's see how this new structure goes in real life. I think the code got more readable after these changes.

playerx committed:
removed commented code

felipesabino commented:
@playerx Thanks a lot for the contribution! Just published [0.7.0](https://github.com/indigotech/graphql-schema-decorator/releases/tag/0.7.0) with these changes
